source/server/listener_impl.cc

```diff
@@ -372,6 +372,32 @@ ListenerImpl::ListenerImpl(ListenerImpl& origin,
   buildProxyProtocolListenerFilter();
   buildTlsInspectorListenerFilter();
   open_connections_ = origin.open_connections_;
+
+#ifdef WIN32
+  // On Windows we need to do some extra validation for the Original Destination filter.
+  // In particular we need to check if:
+  // 1. The platform supports the original destination feature
+  // 2. The `traffic_direction` property is set on the listener. This is required to redirect the
+  // traffic.
+  auto it =
+      std::find_if(config_.listener_filters().begin(), config_.listener_filters().end(),
+                   [](const envoy::config::listener::v3::ListenerFilter& listener) {
+                     return listener.name() ==
+                            Extensions::ListenerFilters::ListenerFilterNames::get().OriginalDst;
+                   });
+  if (it != config_.listener_filters().end()) {
+    if (Network::Win32SupportsOriginalDestination() &&
+        config_.traffic_direction() == envoy::config::core::v3::UNSPECIFIED) {
+      throw EnvoyException("[Windows] Setting original destination filter on a listener without "
+                           "specifying the traffic_direction."
+                           "Configure the traffic_direction listener option");
+    } else {
+      ENVOY_LOG(warn,
```

Comment:
I'm not sure I like the idea of different config being valid on only some versions of windows.I think the above validation should be done whether or not `Win32SupportsOriginalDestination()` is true. 

@htuch Any differing opinion on that? I'm also slightly uncomfortable with a config leaving direction unspecified, which will load and work on non-windows, and not be loadable on windows, but I think it's probably better than the alternative (loading a config we know does not work).

Reply:
I will refactor the if statement here to better reflect what the situation is.

Reply:
The code is now refactored in the following way:

```
if original_dst filter exists:
   if traffic_direction is unspecified:
      throw exception
   
   if windows versions does not support it:
      log warning
```

Reply:
IMHO we should throw if not supported; warnings are a silent failure and not particularly robust for most operators. 

Also, should this be done inside the `original_dst` filter factory config implementation? I think the listener is a strange place to do the validation.

Reply:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/71ca1453e586dbd51f07d08e5efa02563789c97d#diff-bf7c14c67a2a4f213ec7d90ee3a353c7c25381e11d5cec70a40c0093c007dff8R26 regarding where is validation is done.

I did not want to break compatibility this is why I opted for a warning. Given that currently if you specify the `original_dst` filter Envoy loads but does not work. I agree with @htuch that throwing an exception is the best way to fail fast.

@envoyproxy/windows-dev feel free to chime in.

Reply:
Also another thing is that on POSIX if `SOL_IP` is not defined then the user can still use the original destination filter even though it wont work. I am not sure how common this scenario is though